core/src/instance.rs

```diff
@@ -92,41 +99,59 @@ impl Instance {
             loop {
                 match rx_action.recv_timeout(Duration::from_millis(REDUX_LOOP_TIMEOUT_MS)) {
                     Ok(action_wrapper) => {
-                        // Mutate state
-                        {
-                            let mut state = state_mutex.write().unwrap();
-                            *state = state.reduce(
-                                context.clone(),
-                                action_wrapper,
-                                &tx_action,
-                                &tx_observer,
-                            );
-                        }
-
-                        // Add new observers
-                        while let Ok(observer) = rx_observer.try_recv() {
-                            state_observers.push(Box::new(observer));
-                        }
-
-                        // Run all observer closures
-                        {
-                            let state = state_mutex.read().unwrap();
-                            state_observers = state_observers
-                                .into_iter()
-                                .map(|mut observer| {
-                                    observer.check(&state);
-                                    observer
-                                })
-                                .filter(|observer| !observer.done)
-                                .collect::<Vec<_>>();
-                        }
+                        state_observers = sync_self.process_action(
+                            action_wrapper,
+                            state_observers,
+                            &rx_observer,
+                            &context,
+                        );
                     }
                     Err(ref _recv_error) => {}
                 }
             }
         });
     }
 
+    /// Calls the reducers for an action and calls the observers with the new state
+    /// returns the new vector of observers
+    fn process_action(
```

Comment:
can we get some basic tests around this to show that it is compatible with testing now?